File: Weather/Weather/ViewModels/ConditionsViewModel.cs
using System;
using System.ComponentModel;
using Weather.Services;
using Xamarin.Forms;

namespace Weather.ViewModels
{
    class ConditionsViewModel : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler PropertyChanged;

        public ConditionsViewModel()
        {
            NWSService.Service.DataUpdated += NwsService_DataUpdated;
        }

        void NwsService_DataUpdated(object sender, EventArgs e)
        {
            Location = "Current Condition Unavailable";
            TextDescription = null;
            Temperature = null;
            DewPoint = null;
            RelativeHumidity = null;
            Wind = null;
            ConditionsIcon = null;
            Gust = null;
            Visibility = null;
            Pressure = null;
            WindChill = null;
            HeatIndex = null;

            var nwsService = NWSService.Service;

            if (nwsService.IsValid)
            {
                var windDirection = "";
                if (nwsService.WindDirection > 337.5 || nwsService.WindDirection <= 22.5)
                    windDirection = "N";
                else if (nwsService.WindDirection > 22.5 && nwsService.WindDirection <= 67.5)
                    windDirection = "NE";
                else if (nwsService.WindDirection > 67.5 && nwsService.WindDirection <= 112.5)
                    windDirection = "E";
                else if (nwsService.WindDirection > 112.5 && nwsService.WindDirection <= 157.5)
                    windDirection = "SE";
                else if (nwsService.WindDirection > 157.5 && nwsService.WindDirection <= 202.5)
                    windDirection = "S";
                else if (nwsService.WindDirection > 202.5 && nwsService.WindDirection <= 247.5)
                    windDirection = "SW";
                else if (nwsService.WindDirection > 247.5 && nwsService.WindDirection <= 292.5)
                    windDirection = "W";
                else if (nwsService.WindDirection > 292.5 && nwsService.WindDirection <= 337.5)
                    windDirection = "NW";

                // update current condition
                Location = nwsService.Location;
                TextDescription = nwsService.TextDescription;
                Temperature = $"{nwsService.Temperature:F0}℉";
                DewPoint = $"Dew Point {nwsService.DewPoint:F0}℉";
                RelativeHumidity = $"Humidity {nwsService.RelativeHumidity:F0}%";
                Wind = $"Wind {windDirection} {nwsService.WindSpeed:F1} MPH";
                if (!string.IsNullOrEmpty(nwsService.ConditionIconUrl))
                    ConditionsIcon = ImageSource.FromUri(new Uri(nwsService.ConditionIconUrl));

                ShowGust = nwsService.WindGust.HasValue;
                Gust = $"Gust {nwsService.WindGust:F1} MPH";
        
                ShowVisibility = nwsService.Visibility.HasValue;
                Visibility = $"Visibility {nwsService.Visibility:F1} Miles";

                ShowPressure = nwsService.Pressure.HasValue;
                Pressure = $"Pressure {nwsService.Pressure:F2} HG";

                ShowWindChill = nwsService.WindChill.HasValue;
                WindChill = $"Wind Chill {nwsService.WindChill:F0}℉";

                ShowHeatIndex = nwsService.HeatIndex.HasValue;
                HeatIndex = $"Heat Index {nwsService.HeatIndex:F0}℉";
            }

            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Location)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(TextDescription)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Temperature)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(DewPoint)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(RelativeHumidity)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Wind)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ConditionsIcon)));

            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ShowGust)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Gust)));
            
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ShowVisibility)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Visibility)));
            
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ShowPressure)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Pressure)));

            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ShowWindChill)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(WindChill)));

            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ShowHeatIndex)));
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(HeatIndex)));
        }

        public string Location { private set; get; }
        public ImageSource ConditionsIcon { private set; get; }

        public string TextDescription { private set; get; }
        public string Temperature { private set; get; }
        public string Wind { private set; get; }
        public string DewPoint { private set; get; }
        public string RelativeHumidity { private set; get; }
        public bool ShowGust { private set; get; }
        public string Gust { private set; get; }
        
        public bool ShowVisibility { private set; get; }
        public string Visibility { private set; get; }

        public bool ShowPressure { private set; get; }
        public string Pressure { private set; get; }

        public bool ShowWindChill { private set; get; }
        public string WindChill { private set; get; }

        public bool ShowHeatIndex { private set; get; }
        public string HeatIndex { private set; get; }
    }
}

File: native/app/src/main/java/com/companyname/weather/viewModels/ConditionsViewModel.kt
package com.companyname.weather.viewModels

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.companyname.weather.services.NWSService
import java.text.SimpleDateFormat

class ConditionsViewModel : ViewModel() {
    val timestamp: MutableLiveData<String> = MutableLiveData()
    val details: MutableLiveData<DetailedConditionsViewModel.Details> = MutableLiveData()

    init { NWSService.conditions.observeForever { conditions -> onConditionsChanged(conditions) } }

    private fun onConditionsChanged(conditions: NWSService.Conditions?) {
        this.details.postValue(
            conditions?.let { c ->
                DetailedConditionsViewModel.Details(
                    title = "Current Conditions",
                    dewPoint = c.dewPoint?.let { "Dew Point %.0f℉".format(it.toFahrenheit) },
                    relativeHumidity = c.relativeHumidity?.let { "Humidity %.0f%%".format(it) },
                    temperature = c.temperature?.let { "%.0f℉".format(it.toFahrenheit) },
                    wind = getWind(c),
                    windGust = c.windGust?.let { "Gusts %.1f MPH".format(it.toMPH) },
                    icon = c.icon,
                    shortDescription = c.textDescription,
                    visibility = c.visibility?.let { "Visibility %.1f  Miles".format(it.toMiles) },
                    pressure = c.barometricPressure?.let { "Pressure %.2f HG".format(it.toHG) },
                    windChill = c.windChill?.let { "Wind Chill %.0f℉".format(it.toFahrenheit) },
                    heatIndex = c.heatIndex?.let { "Heat Index %.0f℉".format(it.toFahrenheit) }
                )
            } ?: DetailedConditionsViewModel.Details(title = "Forecast not available"))
        this.timestamp.postValue(conditions?.timestamp?.let { "Update ${SimpleDateFormat.getDateTimeInstance().format(it)}" })
    }

    private val Double.toFahrenheit: Double get() = this * 9.0 / 5.0 + 32.0
    private val Double.toMPH: Double get() = this * 2.237
    private val Double.toMiles: Double get() = this / 1609.0
    private val Double.toHG: Double get() = this / 3386.0

    private fun getWind(conditions: NWSService.Conditions):String? {
        conditions.windDirection ?: return null
        conditions.windSpeed ?: return null

        val windDirection = when {
            conditions.windDirection > 337.5 || conditions.windDirection <= 22.5 -> "N"
            conditions.windDirection > 22.5 && conditions.windDirection <= 67.5 -> "NE"
            conditions.windDirection > 67.5 && conditions.windDirection <= 112.5 -> "E"
            conditions.windDirection > 112.5 && conditions.windDirection <= 157.5 -> "SE"
            conditions.windDirection > 157.5 && conditions.windDirection <= 202.5 -> "S"
            conditions.windDirection > 202.5 && conditions.windDirection <= 247.5 -> "SW"
            conditions.windDirection > 247.5 && conditions.windDirection <= 292.5 -> "W"
            conditions.windDirection > 292.5 && conditions.windDirection <= 337.5 -> "NW"
            else -> ""
        }

        return "Wind $windDirection %.1f MPH".format(conditions.windSpeed.toMPH)
    }
}
